databuilder/extractor/dashboard/tableau/tableau_dashboard_utils.py

```diff
@@ -0,0 +1,220 @@
+    EXTERNAL_CLUSTER_NAME = const.EXTERNAL_CLUSTER_NAME
+    EXTERNAL_SCHEMA_NAME = const.EXTERNAL_SCHEMA_NAME
+    CLUSTER = const.CLUSTER
+    DATABASE = const.DATABASE
+
+    @staticmethod
+    def sanitize_schema_name(str):
+        """
+        Sanitizes a given string so that it can safely be used as a table's schema.
+        Sanitization behaves as follows:
+            - all spaces and periods are replaced by underscores
+            - any [], (), -, &, and ? characters are deleted
+        """
+        # type: (str) -> str
+        # this indentation looks a little odd, but otherwise the linter complains
+        return re.sub(r' ', '_',
+                      re.sub(r'\.', '_',
+                             re.sub(r'(\[|\]|\(|\)|\-|\&|\?)', '', str)))
+
+    @staticmethod
+    def sanitize_database_name(str):
+        """
+        Sanitizes a given string so that it can safely be used as a table's database.
+        Sanitization behaves as follows:
+            - all hyphens are deleted
+        """
+        # type: (str) -> str
+        return re.sub(r"-", "", str)
+
+    @staticmethod
+    def sanitize_table_name(str):
+        """
+        Sanitizes a given string so that it can safely be used as a table name.
+        Replicates the current behavior of sanitize_workbook_name, but this is purely coincidental.
+        As more breaking characters/patterns are found, each method should be updated to reflect the specifics.
+        Sanitization behaves as follows:
+            - all forward slashes and single quotes characters are deleted
+        """
+        # type: (str) -> str
+        return re.sub(r'(\/|\')', '', str)
+
+    @staticmethod
+    def sanitize_workbook_name(str):
+        """
+        Sanitizes a given string so that it can safely be used as a workbook ID.
+        Mimics the current behavior of sanitize_table_name for now, but is purely coincidental.
+        As more breaking characters/patterns are found, each method should be updated to reflect the specifics.
+        Sanitization behaves as follows:
+            - all forward slashes and single quotes characters are deleted
+        """
+        # type: (str) -> str
+        return re.sub(r'(\/|\')', '', str)
+
+
+class TableauGraphQLApiExtractor(Extractor):
+    """
+    Base class for querying the Tableau Metdata API, which uses a GraphQL schema.
+    """
+
+    API_VERSION = const.API_VERSION
+    TABLEAU_HOST = const.TABLEAU_HOST
+    SITE_NAME = const.SITE_NAME
+    TABLEAU_ACCESS_TOKEN_NAME = const.TABLEAU_ACCESS_TOKEN_NAME
+    TABLEAU_ACCESS_TOKEN_SECRET = const.TABLEAU_ACCESS_TOKEN_SECRET
+    EXCLUDED_PROJECTS = const.EXCLUDED_PROJECTS
+    EXTERNAL_CLUSTER_NAME = const.EXTERNAL_CLUSTER_NAME
+    EXTERNAL_SCHEMA_NAME = const.EXTERNAL_SCHEMA_NAME
+    EXTERNAL_TABLE_TYPES = const.EXTERNAL_TABLE_TYPES
+    CLUSTER = const.CLUSTER
+    DATABASE = const.DATABASE
+
+    def init(self, conf, auth_token, query, query_variables={}):
```

Comment:
I know we had discussed before that `auth_token, query, query_variables` should be config instead of args here. As it is, this `init` signature does not match `Extractor` parent.